src/llama_index_alloydb_pg/async_reader.py

```diff
@@ -0,0 +1,271 @@
+    """JSON document formatter."""
+    dictionary = {}
+    for column in content_columns:
+        if column in row:
+            dictionary[column] = row[column]
+    return json.dumps(dictionary)
+
+
+def _parse_doc_from_row(
+    content_columns: Iterable[str],
+    metadata_columns: Iterable[str],
+    row: dict,
+    formatter: Callable = text_formatter,
+    metadata_json_column: Optional[str] = DEFAULT_METADATA_COL,
+) -> Document:
+    """Parse row into document."""
+    text = formatter(row, content_columns)
+    metadata: dict[str, Any] = {}
+    # unnest metadata from llamaindex_metadata column
+    if metadata_json_column and row.get(metadata_json_column):
+        for k, v in row[metadata_json_column].items():
+            metadata[k] = v
+    # load metadata from other columns
+    for column in metadata_columns:
+        if column in row and column != metadata_json_column:
+            metadata[column] = row[column]
+
+    return Document(text=text, extra_info=metadata)
+
+
+class AsyncAlloyDBReader(BasePydanticReader):
+    """Load documents from AlloyDB.
+
+    Each document represents one row of the result. The `content_columns` are
+    written into the `text` of the document. The `metadata_columns` are written
+    into the `metadata` of the document. By default, first columns is written into
+    the `text` and everything else into the `metadata`.
+    """
+
+    __create_key = object()
+
+    model_config = ConfigDict(extra="allow")
+
+    def __init__(
+        self,
+        key: object,
+        pool: AsyncEngine,
+        query: str,
+        content_columns: list[str],
+        metadata_columns: list[str],
+        formatter: Callable,
+        metadata_json_column: Optional[str] = None,
+        is_remote: bool = True,
+    ) -> None:
+        """AsyncAlloyDBReader constructor.
+
+        Args:
+            key (object): Prevent direct constructor usage.
+            engine (AlloyDBEngine): AsyncEngine with pool connection to the alloydb database
+            query (Optional[str], optional): SQL query. Defaults to None.
+            content_columns (Optional[list[str]], optional): Column that represent a Document's page_content. Defaults to the first column.
+            metadata_columns (Optional[list[str]], optional): Column(s) that represent a Document's metadata. Defaults to None.
+            formatter (Optional[Callable], optional): A function to format page content (OneOf: format, formatter). Defaults to None.
+            metadata_json_column (Optional[str], optional): Column to store metadata as JSON. Defaults to "llamaindex_metadata".
+            is_remote (bool): Whether the data is loaded from a remote API or a local file.
+
+        Raises:
+            Exception: If called directly by user.
+        """
+        if key != AsyncAlloyDBReader.__create_key:
+            raise Exception("Only create class through 'create' method!")
+
+        super().__init__(is_remote=is_remote)
```

Comment:
With the other changes described in comments, this then should be unnecessary and can be updated to:
```suggestion
        super().__init__()
```